RES-GUENTHER/extra/pipe/pipe02_mit_semaphoren.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * Zweck des Programms:
+ * demonstriert die Funktion von Pipes zur Interprozesskommunikation, indem der Vaterprozess dem Kindprozess eine Konstante MSGKIND ("Chrrr Chrrr").
+ * Mithilfe einer Pipe. Der Kindprozess gibt diese Zeichenkette aus. Anschließend gibt der Kindprozess
+ * Zusätzlich werden Semaphoren verwendet, damit garantiert zuerst geschrieben und dann gelesen wird.
```

Comment:
@RvNovae @ZeroPointMax @nukerxy Sitze mit Luis gerade auch an diesem Beispiel. Uns wird der Sinn der Verwendung von Semaphoren nicht ganz klar. Wenn pipes genutzt werden, wird doch beim Lesen sowieso gewartet, bis ein `EOF` kommt. Wir sehen da keine Möglichkeit für ein Problem im zeitlichen Programmablauf. Kann uns das vielleicht jemand erklären 😅 

Hier noch unser Lösungsentwurf für diese Aufgabe:

```c
/*Zweck: Pipes ausprobieren
	Nachricht von Vater-Prozess zu Kind-Prozess senden.
*/
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <string.h>

int main(){
	int pipepc[2];
	int pipecp[2];

	if(pipe(pipepc) == -1 || pipe(pipecp) == -1){
		printf("pipe error\n");
		exit(EXIT_FAILURE);
	}

	pid_t pid = fork();

	if(pid==-1){
		printf("fork error\n");
		exit(EXIT_FAILURE);
	}

	setbuf(stdout, NULL);

	if(pid == 0){ // Kindprozess
		close(pipepc[1]); // nicht benötigtes Schreib-Ende der PC-Pipe schließen
		close(pipecp[0]); // nicht benötigtes Lese-Ende der CP-Pipe schließen

		// Pipe auslesen -> in Buffer speichern und ausgeben
		char buf;
		while(read(pipepc[0], &buf, 1) > 0) printf("%c", buf);

		char msg[] = "Ich bin dein Sohn\n";
		write(pipecp[1], msg, strlen(msg));

		close(pipepc[0]); // PC-Pipe schließen
		close(pipecp[1]); // CP-Pipe schließen -> EOF marker
	}else{ // Vaterprozess
		close(pipepc[0]); // nicht benötigtes Lese-Ende der PC-Pipe schließen
		close(pipecp[1]); // nicht benötigtes Schreib-Ende der CP-Pipe schließen

		char msg[] = "Ich bin dein Vater\n";
		write(pipepc[1], msg, strlen(msg));
		close(pipepc[1]); // PC-Pipe schließen -> EOF marker

		// Pipe auslesen -> in Buffer speichern und ausgeben
		char buf;
		while(read(pipecp[0], &buf, 1) > 0) printf("%c", buf);

		close(pipepc[0]); // CP-Pipe schließen
	}

	fflush(stdout);
	return 0;
}

```

Reply:
Ach joa das ist meine "komische" Lösung. Wenn man nur eine Pipe hat, wird die Synchronisierung schwierig. Man kann es auch ohne lösen, ja.

Da ich also nur eine Pipe habe, kann ich kein EOF-Marker senden, weil ich die Pipe noch brauche.

Reply:
Um das auszuführen: 
1. Liest du zeichenweise?
a) bis EOF-> write prozess muss close ausführen -> Pipe ist weg, nicht wiederherstellbar: Wird in pipe-02 dann doof. Da soll der Sohn auch eine Nachricht senden
b) terminierendes Symbol definieren, darauf prüfen--> Lesen abbrechen
2. Liest du alles auf einen Rutsch?
Pipe blockiert, wartet, wenn nichts auf der Pipe steht. Read wird einmal ausgeführt, wenn etwas ankommt. Programm läuft weiter. <-> Art des Schreibens relevant (zeichenweise/gesamte Länge)

Die Semaphoren stellen sich, in diesem Beispiel, als überflüssig heraus. Read wartet, bis etwas auf der Pipe steht.
Aber für spätere Aufgaben wird das nützlich!
@Steve2955

Reply:
Wollen wir die Lösung mit den zwei Pipes vielleicht noch hinzufügen? Sieht meiner Ansicht nach auch übersichtlicher aus.

Reply:
@RvNovae gerne!